React native Kotlin examples (#3030)
facebook/react-native-website: docs/native-modules-android.md

We recommend using Android Studio to write your native code. Android studio is an IDE built for Android development and using it will help you resolve minor issues like code syntax errors quickly.

We also recommend enabling [Gradle Daemon](https://docs.gradle.org/2.9/userguide/gradle_daemon.html) to speed up builds as you iterate on Java/Kotlin code.

### Create A Custom Native Module File

> The first step is to create the (CalendarModule.java or CalendarModule.kt) Java/Kotlin file inside `android/app/src/main/java/com/your-app-name/` folder. This Java/Kotlin file will contain your native module Java/Kotlin class.

Reviewer: ```suggestion
The first step is to create the (`CalendarModule.java` or `CalendarModule.kt`) Java/Kotlin file inside `android/app/src/main/java/com/your-app-name/` folder. This Java/Kotlin file will contain your native module Java/Kotlin class.
```

Reviewer: Plus here there could be confusion on the folder as there is Java in this path `android/app/src/main/java/com/your-app-name/`.

Maybe adding a `(the folder is the same for both for either Kotlin of Java)` would help clarify.

}
```

</TabItem>
<TabItem value="kotlin">

```kotlin
import android.util.Log

@ReactMethod
fun createCalendarEvent(name: String, location: String) {
    Log.d("CalendarModule", "Create event called with name: $name and location: $location")
}
```

</TabItem>
</Tabs>

Once you finish implementing the native module and hook it up in JavaScript, you can follow [these steps](https://developer.android.com/studio/debug/am-logcat.html) to view the logs from your app.

> ### Synchronous Methods

Reviewer: There is a snipped here missing.

}
```

</TabItem>
<TabItem value="kotlin">

```kotlin
package com.your-app-name // replace your-app-name with your app’s name

import android.view.View
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ReactShadowNode
import com.facebook.react.uimanager.ViewManager

class MyAppPackage : ReactPackage {

    override fun createViewManagers(reactContext: ReactApplicationContext):
            MutableList<ViewManager<View, ReactShadowNode<*>>> =
        mutableListOf()

    override fun createNativeModules(reactContext: ReactApplicationContext):
            MutableList<NativeModule> {
        val modules = mutableListOf<NativeModule>()
        modules.add(CalendarModule(reactContext))
        return modules

>     }

Reviewer: Maybe simplify to:
```
listOf(CalendarModule(reactContext)).toMutableList()
```

or use 

```
 = mutableListOf().apply { add(CalendarModule(reactContext)) }
```

> To register the `CalendarModule` package, you must add `MyAppPackage` to the list of packages returned in ReactNativeHost's `getPackages()` method. Open up your `MainApplication.java` or `MainApplication.kt` file, which can be found in the following path: `android/app/src/main/java/com/your-app-name/MainApplication.java` or `android/app/src/main/kotlin/com/your-app-name/MainApplication.kt`.

Reviewer: As above, let's stick to a single source folder. 
Given that `android/app/src/main/java/` works for both Java and Kotlin, I'd stick to that.

  }
```

</TabItem>
<TabItem value="kotlin">

```kotlin

> override fun getPackages(): List<ReactPackage> {
>     val packages: MutableList<ReactPackage> = PackageList(this).packages
>     // Packages that cannot be autolinked yet can be added manually here, for example:
>     // packages.add(new MyReactNativePackage());
>     packages.add(MyAppPackage())
>     return packages
> }

Reviewer: ```suggestion
override fun getPackages(): List<ReactPackage> = PackageList(this).packages.apply {
    // Packages that cannot be autolinked yet can be added manually here, for example:
    // add(new MyReactNativePackage());
    add(MyAppPackage())
}
```

  }
```

</TabItem>
<TabItem value="kotlin">

```kotlin
override fun getPackages(): List<ReactPackage> =
    PackageList(this).packages.apply {
        // Packages that cannot be autolinked yet can be added manually here, for example:
        // packages.add(new MyReactNativePackage());
        add(MyAppPackage())
    }

> 

Reviewer 2: ```suggestion
```

Small whitespace cleanup suggestion.

### Argument Types

When a native module method is invoked in JavaScript, React Native converts the arguments from JS objects to their Java/Kotlin object analogues. So for example, if your Java Native Module method accepts a double, in JS you need to call the method with a number. React Native will handle the conversion for you. Below is a list of the argument types supported for native module methods and the JavaScript equivalents they map to.

| Java/Kotlin   | JavaScript |
| ------------- | ---------- |
| Boolean       | ?boolean   |

> | boolean       | boolean    |

Reviewer: Kotlin doesn't primitive types (e.g. `boolean`). 

Maybe `boolean` and `double` in this table should be exported to a separate Java only table?

Reviewer: Same for `int` and `float` below

```

</TabItem>
<TabItem value="kotlin">

```kotlin
    val dateFormat = "yyyy-MM-dd"
    val sdf = SimpleDateFormat(dateFormat, Locale.US)
    val eStartDate = Calendar.getInstance()

>     sdf.parse(startDate)?.let {

Reviewer: Previously there was a `try` there that you removed. I'm assuming it was needed to catch parsing exceptions therefore should be kept here as well (maybe with a `runCatching` or something more idiomatic).

}
```

</TabItem>
<TabItem value="kotlin">

```kotlin

> override fun getConstants(): MutableMap<String, Any> {
>     val constants: MutableMap<String, Any> = HashMap()
>     constants["DEFAULT_EVENT_NAME"] = "New Event"
>     return constants
> }

Reviewer: ```suggestion
override fun getConstants(): MutableMap<String, Any> = hashMapOf(
    "DEFAULT_EVENT_NAME" to "New Event"
)
```

There are two approaches to error handling with callbacks. The first is to follow Node’s convention and treat the first argument passed to the callback as an error object.

<Tabs groupId="android-language" defaultValue={constants.defaultAndroidLanguage} values={constants.androidLanguages}>
<TabItem value="java">

```java
  @ReactMethod
   public void createCalendarEvent(String name, String location, Callback callBack) {
       Integer eventId = ...
       callBack.invoke(null, eventId);
   }
```

</TabItem>
<TabItem value="kotlin">

```kotlin
  @ReactMethod
  fun createCalendarEvent(name: String, location: String, callback: Callback) {
      val eventId = ...

>       callback.invoke(null, eventId)

Reviewer: Nothing to address here, but that's so kotlin sad and unfriendly 😅  @ShikaSD @mdvacca

Another option is to use an onSuccess and onFailure callback:

<Tabs groupId="android-language" defaultValue={constants.defaultAndroidLanguage} values={constants.androidLanguages}>
<TabItem value="java">

```java
@ReactMethod
public void createCalendarEvent(String name, String location, Callback myFailureCallback, Callback mySuccessCallback) {
}
```

</TabItem>
<TabItem value="kotlin">

```kotlin

> @ReactMethod

Reviewer: That's wrongly aligned? All the code should be formatted with 
https://facebookincubator.github.io/ktfmt/

sendEvent(reactContext, "EventReminder", params);
```

</TabItem>
<TabItem value="kotlin">

```kotlin
private fun sendEvent(reactContext: ReactContext, eventName: String, params: WritableMap?) {
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)

> }

Reviewer 2: One more thing. The Java example includes the imports at the beginning of code, do we need those in Kotlin file too?

                    pickerPromise?.let { promise ->
                        when (resultCode) {
                            Activity.RESULT_CANCELED ->
                                promise.reject(E_PICKER_CANCELLED, "Image picker was cancelled")
                            Activity.RESULT_OK -> {
                                val uri = intent?.data

                                uri?.let { promise.resolve(uri.toString()) }
                                    ?: promise.reject(E_NO_IMAGE_DATA_FOUND, "No image data found")
                            }
                        }

                        pickerPromise = null
                    }
                }
            }
        }

    init {
        reactContext.addActivityEventListener(activityEventListener)
    }

    override fun getName() = "ImagePickerModule"

    @ReactMethod
    fun pickImage(promise: Promise) {

        if (currentActivity == null) {
            promise.reject(E_ACTIVITY_DOES_NOT_EXIST, "Activity doesn't exist")
            return
        }

        // Store the promise to resolve/reject when picker returns data
        pickerPromise = promise

        try {

>             val galleryIntent = Intent(Intent.ACTION_PICK)

Reviewer: ```suggestion
            val galleryIntent = Intent(Intent.ACTION_PICK).apply {
                type = "image/\*"
            }
```

            galleryIntent.type = "image/\*"

            val chooserIntent = Intent.createChooser(galleryIntent, "Pick an image")

>             currentActivity!!.startActivityForResult(chooserIntent, IMAGE_PICKER_REQUEST)

Reviewer: Please assign `currentActivity` to a val on line 1036.
As you're null checking at line 1037 you'll be smartcasted to a non-null `Activity` here and you don't need the double bang.

            }
        }

    init {
        reactContext.addActivityEventListener(activityEventListener)
    }

    override fun getName() = "ImagePickerModule"

    @ReactMethod
    fun pickImage(promise: Promise) {
        val activity = currentActivity

        if (activity == null) {
            promise.reject(E_ACTIVITY_DOES_NOT_EXIST, "Activity doesn't exist")
            return
        }

        pickerPromise = promise

        try {
            val galleryIntent = Intent(Intent.ACTION_PICK).apply { type = "image\/*" }

            val chooserIntent = Intent.createChooser(galleryIntent, "Pick an image")

            activity.startActivityForResult(chooserIntent, IMAGE_PICKER_REQUEST)
        } catch (t: Throwable) {
            pickerPromise?.reject(E_FAILED_TO_SHOW_PICKER, t)
            pickerPromise = null
        }
    }
}

    companion object {
        const val IMAGE_PICKER_REQUEST = 1
        const val E_ACTIVITY_DOES_NOT_EXIST = "E_ACTIVITY_DOES_NOT_EXIST"
        const val E_PICKER_CANCELLED = "E_PICKER_CANCELLED"
        const val E_FAILED_TO_SHOW_PICKER = "E_FAILED_TO_SHOW_PICKER"
        const val E_NO_IMAGE_DATA_FOUND = "E_NO_IMAGE_DATA_FOUND"
    }

> }

Reviewer 2: Hmm, it looks like there is an additional closing bracket (`}`), which leads to the error, when trying to format the code in `ktfmt` tool.

    @ReactMethod
    fun pickImage(promise: Promise) {

>         val currentActivity = currentActivity

Reviewer: ```suggestion
        val activity = currentActivity
```

Don't name shadow otherwise the IDE will warn users here.
